### ISSUE TITLE
Qiling framework simulates the execution of clang-compiled C++ Mini Program and gets stuck.

### ISSUE THREAD

**Issue by cjlovefree** [qilingframework/qiling#1365]
***Describe the bug**
The qiling framework simulates the problem of unlimited memory reading and writing when initializing various libraries when running a linux Mini Program. 
This program is compiled by clang written by C++. How should this bug be repaired?

**Sample Code**
```python
from qiling import Qiling
from qiling.const import QL_VERBOSE
rootfs = 'x8664_linux'
if __name__ == "__main__":
    ql = Qiling(['easy_RE'], rootfs, verbose=QL_VERBOSE.DEFAULT,profile='linux.ql')
    ql.run()
```

**Screenshots**
![image](https://github.com/qilingframework/qiling/assets/121709972/5f964f5f-6c9e-4713-92b3-8796441e96ae)

**Additional context**
My qiling framework is 1.4.5 currently up to date, and the unicorn framework is 2.0.1post and up to date.

I packaged the defective program and the qiling simulation script in the attachment test.zip. 
Please point out what is wrong with this and how to solve it?
[test.zip](https://github.com/qilingframework/qiling/files/12060834/test.zip)

**Comment by elicn** [comment]
The problem you are describing is unclear. Can you rephrase?
Specifying the actual behavior vs. the expected one would help too.

**Comment by cjlovefree** [comment]
> The problem you are describing is unclear. Can you rephrase? Specifying the actual behavior vs. the expected one would help too.

I use the qiling framework to run the ordinary program in the attachment test.zip above, but the program keeps jamming and cpu takes up a lot of time. I don't know exactly what caused it. It looks like a problem caused by frequent memory allocation. 
The programs in the above attachment test.zip will not get stuck with ida debugging and loading, so you can run it and have a look.

**Comment by cjlovefree** [comment]
> The problem you are describing is unclear. Can you rephrase? Specifying the actual behavior vs. the expected one would help too.
In addition, I would like to add that the final error message is as follows:
[=] 	mmap(addr = 0x0, length = 0x7801000, prot = 0x3, flags = 0x22, fd = 0xffffffff, pgoffset = 0x0) = 0x7fffbfee2000
[=] 	munmap(addr = 0x7fffbc2e1000, length = 0x3c01000) = 0x0
[=] 	mmap(addr = 0x0, length = 0xf001000, prot = 0x3, flags = 0x22, fd = 0xffffffff, pgoffset = 0x0) = 0x7fffc76e3000
[=] 	munmap(addr = 0x7fffbfee2000, length = 0x7801000) = 0x0
[=] 	mmap(addr = 0x0, length = 0x1e001000, prot = 0x3, flags = 0x22, fd = 0xffffffff, pgoffset = 0x0) = 0x7fffd66e4000
[=] 	munmap(addr = 0x7fffc76e3000, length = 0xf001000) = 0x0
[=] 	mmap(addr = 0x0, length = 0x3c001000, prot = 0x3, flags = 0x22, fd = 0xffffffff, pgoffset = 0x0) = 0x80000000e000
[=] 	munmap(addr = 0x7fffd66e4000, length = 0x1e001000) = 0x0
[=] 	mmap(addr = 0x0, length = 0x78001000, prot = 0x3, flags = 0x22, fd = 0xffffffff, pgoffset = 0x0) = 0x80003c00f000
[=] 	munmap(addr = 0x80000000e000, length = 0x3c001000) = 0x0
[x] 	Syscall ERROR: ql_syscall_mmap DEBUG: Invalid argument (UC_ERR_ARG)
Traceback (most recent call last):
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/posix.py", line 280, in load_syscall
    retval = syscall_hook(self.ql, *params)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/syscall/mman.py", line 220, in ql_syscall_mmap
    return syscall_mmap_impl(ql, addr, length, prot, flags, fd, pgoffset, 1)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/syscall/mman.py", line 199, in syscall_mmap_impl
    ql.mem.write(addr, data)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/memory.py", line 353, in write
    self.ql.uc.mem_write(addr, data)
  File "/home/cs/.local/lib/python3.10/site-packages/unicorn/unicorn.py", line 586, in mem_write
    raise UcError(status)
unicorn.unicorn.UcError: Invalid argument (UC_ERR_ARG)
[x] 	CPU Context:
[x] 	ah	: 0x0
[x] 	al	: 0x9
[x] 	ch	: 0x0
[x] 	cl	: 0x22
[x] 	dh	: 0x0
[x] 	dl	: 0x3
[x] 	bh	: 0x0
[x] 	bl	: 0x0
[x] 	ax	: 0x9
[x] 	cx	: 0x22
[x] 	dx	: 0x3
[x] 	bx	: 0x0
[x] 	sp	: 0xd7f8
[x] 	bp	: 0x0
[x] 	si	: 0x1000
[x] 	di	: 0x0
[x] 	ip	: 0xba13
[x] 	eax	: 0x9
[x] 	ecx	: 0x22
[x] 	edx	: 0x3
[x] 	ebx	: 0x0
[x] 	esp	: 0xd7f8
[x] 	ebp	: 0x0
[x] 	esi	: 0xf0001000
[x] 	edi	: 0x0
[x] 	eip	: 0xb843ba13
[x] 	rax	: 0x9
[x] 	rbx	: 0x0
[x] 	rcx	: 0x22
[x] 	rdx	: 0x3
[x] 	rsi	: 0xf0001000
[x] 	rdi	: 0x0
[x] 	rbp	: 0x0
[x] 	rsp	: 0x80000000d7f8
[x] 	r8	: 0xffffffff
[x] 	r9	: 0x0
[x] 	r10	: 0x22
[x] 	r11	: 0x0
[x] 	r12	: 0x3
[x] 	r13	: 0xf0001000
[x] 	r14	: 0x22
[x] 	r15	: 0xffffffff
[x] 	rip	: 0x7fffb843ba13
[x] 	cr0	: 0x11
[x] 	cr1	: 0x0
[x] 	cr2	: 0x0
[x] 	cr3	: 0x0
[x] 	cr4	: 0x0
[x] 	cr8	: 0x0
[x] 	dr0	: 0x0
[x] 	dr1	: 0x0
[x] 	dr2	: 0x0
[x] 	dr3	: 0x0
[x] 	dr4	: 0x0
[x] 	dr5	: 0x0
[x] 	dr6	: 0xffff0ff0
[x] 	dr7	: 0x400
[x] 	st0	: 0x0
[x] 	st1	: 0x0
[x] 	st2	: 0x0
[x] 	st3	: 0x0
[x] 	st4	: 0x0
[x] 	st5	: 0x0
[x] 	st6	: 0x0
[x] 	st7	: 0x0
[x] 	eflags	: 0x46
[x] 	cs	: 0x33
[x] 	ss	: 0x28
[x] 	ds	: 0x0
[x] 	es	: 0x0
[x] 	fs	: 0x0
[x] 	gs	: 0x0
[x] 	r8b	: 0xff
[x] 	r9b	: 0x0
[x] 	r10b	: 0x22
[x] 	r11b	: 0x0
[x] 	r12b	: 0x3
[x] 	r13b	: 0x0
[x] 	r14b	: 0x22
[x] 	r15b	: 0xff
[x] 	r8w	: 0xffff
[x] 	r9w	: 0x0
[x] 	r10w	: 0x22
[x] 	r11w	: 0x0
[x] 	r12w	: 0x3
[x] 	r13w	: 0x1000
[x] 	r14w	: 0x22
[x] 	r15w	: 0xffff
[x] 	r8d	: 0xffffffff
[x] 	r9d	: 0x0
[x] 	r10d	: 0x22
[x] 	r11d	: 0x0
[x] 	r12d	: 0x3
[x] 	r13d	: 0xf0001000
[x] 	r14d	: 0x22
[x] 	r15d	: 0xffffffff
[x] 	fsbase	: 0x7fffb8711b80
[x] 	gsbase	: 0x6000000
[x] 	xmm0	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm1	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm2	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm3	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm4	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm5	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm6	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm7	: 0xffffffffffffffffffffffffffffffff
[x] 	xmm8	: 0x7fffb870c420
[x] 	xmm9	: 0x0
[x] 	xmm10	: 0x0
[x] 	xmm11	: 0x0
[x] 	xmm12	: 0x0
[x] 	xmm13	: 0x0
[x] 	xmm14	: 0x0
[x] 	xmm15	: 0x0
[x] 	xmm16	: 0x0
[x] 	xmm17	: 0x0
[x] 	xmm18	: 0x0
[x] 	xmm19	: 0x0
[x] 	xmm20	: 0x0
[x] 	xmm21	: 0x0
[x] 	xmm22	: 0x0
[x] 	xmm23	: 0x0
[x] 	xmm24	: 0x0
[x] 	xmm25	: 0x0
[x] 	xmm26	: 0x0
[x] 	xmm27	: 0x0
[x] 	xmm28	: 0x0
[x] 	xmm29	: 0x0
[x] 	xmm30	: 0x0
[x] 	xmm31	: 0x0
[x] 	ymm0	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm1	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm2	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm3	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm4	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm5	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm6	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm7	: 0xffffffffffffffffffffffffffffffff
[x] 	ymm8	: 0x7fffb870c42000007fffb870c420
[x] 	ymm9	: 0x0
[x] 	ymm10	: 0x0
[x] 	ymm11	: 0x0
[x] 	ymm12	: 0x0
[x] 	ymm13	: 0x0
[x] 	ymm14	: 0x0
[x] 	ymm15	: 0x0
[x] 	ymm16	: 0x0
[x] 	ymm17	: 0x0
[x] 	ymm18	: 0x0
[x] 	ymm19	: 0x0
[x] 	ymm20	: 0x0
[x] 	ymm21	: 0x0
[x] 	ymm22	: 0x0
[x] 	ymm23	: 0x0
[x] 	ymm24	: 0x0
[x] 	ymm25	: 0x0
[x] 	ymm26	: 0x0
[x] 	ymm27	: 0x0
[x] 	ymm28	: 0x0
[x] 	ymm29	: 0x0
[x] 	ymm30	: 0x0
[x] 	ymm31	: 0x0
[x] 	zmm0	: 0x0
[x] 	zmm1	: 0x0
[x] 	zmm2	: 0x0
[x] 	zmm3	: 0x0
[x] 	zmm4	: 0x0
[x] 	zmm5	: 0x0
[x] 	zmm6	: 0x0
[x] 	zmm7	: 0x0
[x] 	zmm8	: 0x0
[x] 	zmm9	: 0x0
[x] 	zmm10	: 0x0
[x] 	zmm11	: 0x0
[x] 	zmm12	: 0x0
[x] 	zmm13	: 0x0
[x] 	zmm14	: 0x0
[x] 	zmm15	: 0x0
[x] 	zmm16	: 0x0
[x] 	zmm17	: 0x0
[x] 	zmm18	: 0x0
[x] 	zmm19	: 0x0
[x] 	zmm20	: 0x0
[x] 	zmm21	: 0x0
[x] 	zmm22	: 0x0
[x] 	zmm23	: 0x0
[x] 	zmm24	: 0x0
[x] 	zmm25	: 0x0
[x] 	zmm26	: 0x0
[x] 	zmm27	: 0x0
[x] 	zmm28	: 0x0
[x] 	zmm29	: 0x0
[x] 	zmm30	: 0x0
[x] 	zmm31	: 0x0
[x] 	Hexdump:
[x] 	48 3d 00 f0 ff ff 77 7d
[x] 	Disassembly:
[=] 	00007fffb843ba13 [[mmap] libc.so.6     + 0x11ba13]  48 3d 00 f0 ff ff    cmp                  rax, -0x1000
[=] 	00007fffb843ba19 [[mmap] libc.so.6     + 0x11ba19]  77 7d                ja                   0x7fffb843ba98
[=] 	00007fffb843ba1b [[mmap] libc.so.6     + 0x11ba1b]  5b                   pop                  rbx
[=] 	00007fffb843ba1c [[mmap] libc.so.6     + 0x11ba1c]  5d                   pop                  rbp
[=] 	00007fffb843ba1d [[mmap] libc.so.6     + 0x11ba1d]  41 5c                pop                  r12
[=] 	00007fffb843ba1f [[mmap] libc.so.6     + 0x11ba1f]  41 5d                pop                  r13
[=] 	00007fffb843ba21 [[mmap] libc.so.6     + 0x11ba21]  41 5e                pop                  r14
[=] 	00007fffb843ba23 [[mmap] libc.so.6     + 0x11ba23]  41 5f                pop                  r15
[=] 	00007fffb843ba25 [[mmap] libc.so.6     + 0x11ba25]  c3                   ret                  
[=] 	00007fffb843ba26 [[mmap] libc.so.6     + 0x11ba26]  66 2e 0f 1f 84 00 00 00 00 00 nop                  word ptr cs:[rax + rax]
[=] 	00007fffb843ba30 [[mmap] libc.so.6     + 0x11ba30]  48 8b 05 31 f4 2c 00 mov                  rax, qword ptr [rip + 0x2cf431]
[=] 	00007fffb843ba37 [[mmap] libc.so.6     + 0x11ba37]  64 c7 00 16 00 00 00 mov                  dword ptr fs:[rax], 0x16
[=] 	00007fffb843ba3e [[mmap] libc.so.6     + 0x11ba3e]  48 c7 c0 ff ff ff ff mov                  rax, 0xffffffffffffffff
[=] 	00007fffb843ba45 [[mmap] libc.so.6     + 0x11ba45]  c3                   ret                  
[=] 	00007fffb843ba46 [[mmap] libc.so.6     + 0x11ba46]  66 2e 0f 1f 84 00 00 00 00 00 nop                  word ptr cs:[rax + rax]
[=] 	00007fffb843ba50 [[mmap] libc.so.6     + 0x11ba50]  f6 c2 04             test                 dl, 4
[x] 	PC = 0x00007fffb843ba13

[x] 	Memory map:
[x] 	Start                End                  Perm    Label                   Image
[x] 	000000000000030000 - 000000000000031000   rwx     [GDT]                   
[x] 	000000000000400000 - 00000000000042a000   r-x     easy_RE                 /home/cs/cj_android/testqiling/easy_RE
[x] 	000000000000629000 - 00000000000062a000   r--     easy_RE                 /home/cs/cj_android/testqiling/easy_RE
[x] 	00000000000062a000 - 00000000000062b000   rw-     easy_RE                 /home/cs/cj_android/testqiling/easy_RE
[x] 	00000000000062b000 - 00000000000062d000   rwx     [hook_mem]              
[x] 	00000000000062d000 - 00000000000064e000   rwx     [brk]                   
[x] 	00000000000064e000 - 000000000000660000   rwx     [brk]                   
[x] 	0000007fffb7dd6000 - 0000007fffb7e6c000   r--     [mmap] libstdc++.so.6   
[x] 	0000007fffb7e6c000 - 0000007fffb7f5c000   r-x     [mmap] libstdc++.so.6   
[x] 	0000007fffb7f5c000 - 0000007fffb7fa5000   r--     [mmap] libstdc++.so.6   
[x] 	0000007fffb7fa5000 - 0000007fffb7fa6000   ---     [mmap] libstdc++.so.6   
[x] 	0000007fffb7fa6000 - 0000007fffb7fb1000   r--     [mmap] libstdc++.so.6   
[x] 	0000007fffb7fb1000 - 0000007fffb7fb4000   rw-     [mmap] libstdc++.so.6   
[x] 	0000007fffb7fb4000 - 0000007fffb7fb7000   rw-     [mmap anonymous]        
[x] 	0000007fffb7fb7000 - 0000007fffb7fb9000   rw-     [mmap anonymous]        
[x] 	0000007fffb7fb9000 - 0000007fffb7fc8000   r--     [mmap] libm.so.6        
[x] 	0000007fffb7fc8000 - 0000007fffb806f000   r-x     [mmap] libm.so.6        
[x] 	0000007fffb806f000 - 0000007fffb8106000   r--     [mmap] libm.so.6        
[x] 	0000007fffb8106000 - 0000007fffb8107000   r--     [mmap] libm.so.6        
[x] 	0000007fffb8107000 - 0000007fffb8108000   rw-     [mmap] libm.so.6        
[x] 	0000007fffb8108000 - 0000007fffb811f000   r-x     [mmap] libgcc_s.so.1    
[x] 	0000007fffb811f000 - 0000007fffb831e000   ---     [mmap] libgcc_s.so.1    
[x] 	0000007fffb831e000 - 0000007fffb831f000   r--     [mmap] libgcc_s.so.1    
[x] 	0000007fffb831f000 - 0000007fffb8320000   rw-     [mmap] libgcc_s.so.1    
[x] 	0000007fffb8320000 - 0000007fffb8507000   r-x     [mmap] libc.so.6        
[x] 	0000007fffb8507000 - 0000007fffb8707000   ---     [mmap] libc.so.6        
[x] 	0000007fffb8707000 - 0000007fffb870b000   r--     [mmap] libc.so.6        
[x] 	0000007fffb870b000 - 0000007fffb870d000   rw-     [mmap] libc.so.6        
[x] 	0000007fffb870d000 - 0000007fffb8711000   rw-     [mmap anonymous]        
[x] 	0000007fffb8711000 - 0000007fffb8713000   rw-     [mmap anonymous]        
[x] 	0000007fffb8713000 - 0000007fffb8715000   rw-     [mmap anonymous]        
[x] 	0000007ffff7dd5000 - 0000007ffff7dfc000   r-x     ld-linux-x86-64.so.2    /home/cs/cj_android/testqiling/x8664_linux/lib64/ld-linux-x86-64.so.2
[x] 	0000007ffff7ffc000 - 0000007ffff7ffd000   r--     ld-linux-x86-64.so.2    /home/cs/cj_android/testqiling/x8664_linux/lib64/ld-linux-x86-64.so.2
[x] 	0000007ffff7ffd000 - 0000007ffff7fff000   rw-     ld-linux-x86-64.so.2    /home/cs/cj_android/testqiling/x8664_linux/lib64/ld-linux-x86-64.so.2
[x] 	0000007ffffffde000 - 00000080000000e000   rwx     [stack]                 
[x] 	00000080003c00f000 - 0000008000b4010000   rw-     [mmap anonymous]        
[x] 	0000008000b4010000 - 0000008001a4011000   rwx     [mmap anonymous]        
[x] 	00ffffffffff600000 - 00ffffffffff601000   rwx     [vsyscall]              
Traceback (most recent call last):
  File "/home/cs/cj_android/testqiling/qlgetenv.py", line 18, in <module>
    ql.run()
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/core.py", line 582, in run
    self.os.run()
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/linux/linux.py", line 176, in run
    self.ql.emu_start(self.ql.loader.elf_entry, self.exit_point, self.ql.timeout, self.ql.count)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/core.py", line 757, in emu_start
    raise self.internal_exception
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/core_hooks.py", line 94, in wrapper
    return callback(*args, **kwargs)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/core_hooks.py", line 158, in _hook_insn_cb
    ret = hook.call(ql, *hook_args)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/core_hooks_types.py", line 25, in call
    return self.callback(ql, *args)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/linux/linux.py", line 132, in hook_syscall
    return self.load_syscall()
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/posix.py", line 298, in load_syscall
    raise e
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/posix.py", line 280, in load_syscall
    retval = syscall_hook(self.ql, *params)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/syscall/mman.py", line 220, in ql_syscall_mmap
    return syscall_mmap_impl(ql, addr, length, prot, flags, fd, pgoffset, 1)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/posix/syscall/mman.py", line 199, in syscall_mmap_impl
    ql.mem.write(addr, data)
  File "/home/cs/.local/lib/python3.10/site-packages/qiling/os/memory.py", line 353, in write
    self.ql.uc.mem_write(addr, data)
  File "/home/cs/.local/lib/python3.10/site-packages/unicorn/unicorn.py", line 586, in mem_write
    raise UcError(status)
unicorn.unicorn.UcError: Invalid argument (UC_ERR_ARG)

**Comment by elicn** [comment]
I am not sure why, but emulating this binary takes forever. Couldn't figure out why it takes `ld` so long to load.. Anyway, I didn't hit any exception for the time I waited.

**Comment by cjlovefree** [comment]
> I am not sure why, but emulating this binary takes forever. Couldn't figure out why it takes `ld` so long to load.. Anyway, I didn't hit any exception for the time I waited.

Thank you. That's the situation. 
I don't know why it took so long, but at first I thought qiling, the C++ program compiled by golang and Gmail +, didn't support it. 
However, this is obviously not the case. 
After loading this program on a mainframe computer for more than 20 minutes, I got the above exception error.

**Comment by cjlovefree** [comment]
> I emulated the program while creating a heatmap of the executed instruction, to identify the loop we are stuck in. It appears that the program reads the "secret.dump" file char by char until eof. However, the file wasn't opened properly (that is the actual bug) and the program never bothered to check for open errors, so that loop went on forever.
> 
> I can upload a patch in a day or two, but this is how you can work around it in the meantime:
> 
> 1. Go to `qiling/os/posix/syscall/unistd.py` and patch line 216 to be: `dirfd = __as_signed(dirfd, 32)` instead of `dirfd = __as_signed(dirfd, ql.arch.pointersize)`
> 2. Copy both `easy_RE` and `secret.dump` to the emulation rootfs directory, for example: `./examples/rootfs/x8664_linux/`. Make sure both files are on the _root_ of that rootfs
> 
> Now everything should run correctly.

I see. Thank you very much. 
This is a problem that has been bothering me for several days.
